File: frameworks/domain.core/src/main/java/com/example/frameworks/domain/core/ObjectFieldType.java
package com.example.frameworks.domain.core;

import java.util.*;

public class ObjectFieldType extends FieldType<Payload> {
    public static ObjectFieldType asObject() {
        return new ObjectFieldType();
    }

    public ObjectFieldType nullToEmpty() {
        setNullHandler(() -> new Payload(Collections.emptyMap()));
        return this;
    }

    @Override
    protected boolean match(Class<?> underlyingType) {
        return Map.class.isAssignableFrom(underlyingType);
    }

    @Override
    protected Payload convert(Object value) {
        return new Payload(value);
    }
}

File: frameworks/domain.core/src/main/kotlin/com/example/frameworks/domain/core/ObjectFieldType.kt
package com.example.frameworks.domain.core

class ObjectFieldType : FieldType<Payload>() {
    fun nullToEmpty(): ObjectFieldType {
        setNullHandler { Payload(emptyMap<Any, Any>()) }
        return this
    }

    override fun match(underlyingType: Class<*>): Boolean {
        return MutableMap::class.java.isAssignableFrom(underlyingType)
    }

    override fun convert(value: Any): Payload {
        return Payload(value)
    }

    companion object {
        @JvmStatic
        fun asObject(): ObjectFieldType {
            return ObjectFieldType()
        }
    }
}
